### DISCUSSION TITLE
Some differences in the Dashboard that don't make sense when compared to electricity provider information.

### DISCUSSION THREAD

**Discussion by Stephen2615** [jasonacox/Powerwall-Dashboard#542] [General]
Greetings,

Now that the installers have finally fixed their dodgy wiring (thanks to the Dashboard for giving me the information to determine that), I now have what looks like accurate data shown on the Dashboard. There are some significant differences between what my electricity providers data (what I get billed on) and what is shown on the Dashboard. 

I assume that the data comes from the Telsa API and it is much the same format as the data.csv we can download from the Tesla App, which is:

DateTime      Home (kW)	Powerwall (kW)	Solar (kW)	Grid (kW)

The Dashboard shows: 

"To Grid" with 14.1 kWh where my provider shows 28.64 kWh. has been exported. 

"Solar Energy" (48.4 - provider v 48.9 - Tesla) and "From grid" (5.16 - provider v 4.2 - Tesla) which are close enough and are probably just rounding errors for five minute intervals.

The Dashboard has Home Usage at 36.8 kWh. The difference between what is exported to the grid  which is 28.64 kWh and our home usage (charging the battery from our solar and whatever we use of our own solar or battery) which is 48.4 kWh is 19.76 kWh which is what I would think is home usage. 

"Too Powerwall" looks about right at close to 10 kWh over the 24 hour period given there are two distinct charging times (some off peak and the rest solar during the day). 

So I am a bit baffled at "Home Usage" and "To Grid".  This also causes problems with "Net Grid Usage". The Dashboard shows -9.9 kWh where it is 28.64 - 5.6 which is a net grid usage of - 23.04 kWh. 

Am I missing something here?

Cheers

Stephen

![image](https://github.com/user-attachments/assets/c9603db9-c2e2-4559-b0c1-4e12bf86f98f)

**Comment by jasonacox** [comment]
Hi @Stephen2615 - I'm not sure I can reconcile this for you but some ideas:

* The Power Meters data on the dashboard are based on the timeframe and timezone selected. I have seen utility providers lag a bit or not measure same start/end for things like "yesterday".  The better option is to look at a week or month aggregates to compare.
* The meters are simple.  It is just the inegral of all the samples over a given time period. That data is what the gateway (or Tesla API) is reporting for a point in time. 
* You can check the math in the CQ query here: https://github.com/jasonacox/Powerwall-Dashboard/blob/main/influxdb/influxdb.sql and the line that gets the instant W power (used for the summation over time to get kWh) is:

```sql
CREATE CONTINUOUS QUERY cq_autogen ON powerwall
BEGIN
  SELECT
    mean(home) AS home,
    mean(solar) AS solar,
    mean(from_pw) AS from_pw,
    mean(to_pw) AS to_pw,
    mean(from_grid) AS from_grid,
    mean(to_grid) AS to_grid,
    last(percentage) AS percentage
  INTO powerwall.autogen.:MEASUREMENT
  FROM (
    SELECT
      load_instant_power AS home,
      solar_instant_power AS solar,
      abs(
        (1 + battery_instant_power / abs(battery_instant_power)) * battery_instant_power / 2
      ) AS from_pw,
      abs(
        (1 - battery_instant_power / abs(battery_instant_power)) * battery_instant_power / 2
      ) AS to_pw,
      abs(
        (1 + site_instant_power / abs(site_instant_power)) * site_instant_power / 2
      ) AS from_grid,
      abs(
        (1 - site_instant_power / abs(site_instant_power)) * site_instant_power / 2
      ) AS to_grid,
      percentage
    FROM raw.http
  )
  GROUP BY time(1m), month, year
  FILL(linear)
END
```

The math for the summation (integral) of the data from W samples in kWh is here:

```sql
CREATE CONTINUOUS QUERY cq_kwh ON powerwall
RESAMPLE EVERY 1m
BEGIN
  SELECT
    integral(home) / 1000 / 3600 AS home_kwh,
    integral(solar) / 1000 / 3600 AS solar_kwh,
    integral(from_pw) / 1000 / 3600 AS from_pw_kwh,
    integral(to_pw) / 1000 / 3600 AS to_pw_kwh,
    integral(from_grid) / 1000 / 3600 AS from_grid_kwh,
    integral(to_grid) / 1000 / 3600 AS to_grid_kwh
  INTO powerwall.kwh.:MEASUREMENT
  FROM autogen.http
  GROUP BY time(1h), month, year
  TZ('America/Los_Angeles')
END
```

**Reply by hulkster** [reply]
FYI FWIW that I just updated my October numbers.

Powerwall says the "net" (solar production - usage) was 249kWh ... Xcel says it was 247.
For the year so far, Powerwall says 2,892kWh versus 2,2895kWh from Xcel.

So at least in my case, I would say that is pretty darn accurate!    ;-)

P.S. And yes, my "Nega-Watt" strategy (minimize usage) works too well as I'm generating excess energy ... ;-)

**Reply by Stephen2615** [reply]
@hulkster I have been running the Dashboard for ages and I only recently discovered (not paying attention to the numbers) that the installers had incorrectly wired the installation (they used a US instance and not a EU/AU wiring system) so all my long term data is useless. I need to get some more longer term data to properly check. I have 8.5 kWh panels on my roof which on a good summers day creates about 60 kWh. It's almost an obsession with solar and grid reliance.

**Reply by hulkster** [reply]
Bummer they messed this up for you @Stephen2615 - stuff happens. 

I glanced quickly at your numbers ... is there a chance you can somehow multiply by two (or 230/120 volts) to get "close" to the correct historical numbers? I.e. maybe clear the "bad" data and run tesla-history.py (after hacking it a bit) to repopulate InfluxDB with "good" data?

P.S. I was just sharing my data as an example where the agreement is almost exact - about 0.1% difference year-to-date ... which I'm surprised/impressed that it's that close. So good job by the clamp-on ammeters to feed data to the Tesla system which is snarfed & totaled by Powerwall.

**Comment by Stephen2615** [comment]
Hi @jasonacox, thanks for taking the time to provide the information. I will attempt the queries soon. I did notice that our electricity provider doesn't make any daylight savings adjustments with their data but that won't make such a major difference. Just out of curiosity, how often does the Dashboard query whatever it queries?

**Reply by jasonacox** [reply]
The dashboard polls power metrics (W) every 5 seconds which is what will show in "Live Monitoring" and the animation. The "Meters" on the dashboard are computing the integral (kWh) on 15m intervals.

**Comment by Stephen2615** [comment]
@jasonacox 

Hi Jason, 
I have been stuffing about looking at data and I found an entry in the Home Assistant Tesla Powerwall integration called Load Import. Oddly enough, it seems to make the Powerwall think it is consuming more power than it is (even though in reality, it doesn't affect the grid). Your pypowerwall information (awesome as always) mentions "home" as being an alias for "load". I get the feeling that someone (with some connection to the government, be it an installer or a virtual power plant company) has done something that confuses the Tesla but doesn't impact it. I cannot see the advanced section of the Tesla App so it makes me wonder about some setting that I cant see. 

The Home Usage is way too high but if the load is the home usage, it isn't. 

Does that make any sense? 

Cheers

Stephen
